pkg/operator/revisioncontroller/revision_controller.go

```diff
@@ -100,6 +107,16 @@ func (c RevisionController) createRevisionIfNeeded(ctx context.Context, recorder
 	if requiredIsNotFound {
 		err = fmt.Errorf("%v", reason)
 	} else {
+		if c.revisionPrecondition != nil {
+			shouldCreateNewRevision, err := c.revisionPrecondition(ctx)
+			if err != nil {
+				return true, err
+			}
+			if !shouldCreateNewRevision {
+				klog.Infof("Preconditions not met, requeuing the creation of new revison %d triggered by %q", nextRevision, reason)
+				return true, nil
```

Comment:
I am confused here on the difference between syncError and synthetic error, and how factory controller would behave differently in these two cases. Could you please shed some light on that so I can use it to check the questions.

As for the error returned by the precondition check, I believe a an error wouldn’t normally be returned if the precondition evaluates to false. Instead, a an error would only be returned if one is encountered while trying to determine/compute the precondition.

Reply:
> 1. Will the operator go into a degraded state?
Yes, the operator goes to degraded state. Etcd static pod manifest was removed to bring the system to quorum unsafe condition. This is causing the function to requeue here:
```
shouldCreateNewRevision, err := c.revisionPrecondition(ctx)
			if err != nil {
				return true, err
			}

Reply:
Thank you for the feedback.

> 1. Will the operator go into a degraded state?

Yes, the `RevisionController` goes into a degraded state. To verify this, etcd static pod manifest was removed to bring the cluster to quorum-unsafe state. Observing the condition populated in `etcd/cluster,` the following code flow was traced for this case. The `createRevisionIfNeeded` function requeues from here:
https://github.com/openshift/library-go/blob/b43f719e8c7ef2ee471af10cc29218593ee35de8/pkg/operator/revisioncontroller/revision_controller.go#L111-L114

The controller enters degraded state from here:
https://github.com/openshift/library-go/blob/b43f719e8c7ef2ee471af10cc29218593ee35de8/pkg/operator/revisioncontroller/revision_controller.go#L373-L381

> 2. Will the error trigger the execution of the sync method in a backoff manner?

The `sync` method gets called multiple times while the precondition waits for quorum restoration. This can be seen from the logs. However, I am unsure if it is called in a backoff manner.

> 3. If yes, then I believe there is a possibility of delay. Could you check how long it might be?

If this is referring to the time required for quorum restoration, I believe there isn't a way to know this. It could be seconds to infinity (you could theoretically run forever without quorum as long as you weren't writing to the database).

> I'm wondering if we shouldn't change the behaviour in the case of syntheticError.

Form my observations, the current behavior indicates that the `RevisionController` doesn't degrade when the quorum is unsafe. Only the quorum dependant controllers degrade.  However,  after moving the quorum check from the quorum dependant controllers to `RevisionController`, the behavior changes, and now the `RevisionController` degrades when quorum is unsafe while the quorum dependent controller do not. 

I'd appreciate your thoughts on how to not deviate the behaviour in the case of syntheticError. I'm still unsure if I may be missing something from your comments.

Reply:
> Yes, the RevisionController goes into a degraded state. To verify this, etcd static pod manifest was removed to bring the cluster to quorum-unsafe state. Observing the condition populated in etcd/cluster, the following code flow was traced for this case. The createRevisionIfNeeded function requeues from here:

I'm okay with going into a degraded state when an error is returned from the precondition function. I wanted to confirm if the operator will degrade when we return at line 117.

I have checked this, and it seems the operator will not enter a degraded state as long as [WithSyncDegradedOnError](https://github.com/openshift/library-go/blob/e536960980dd4d3e09b13ebc372b0d7d29b7dd75/pkg/controller/factory/factory.go#L235) is not specified (which is the case for the revision controller)

Reply:
> The sync method gets called multiple times while the precondition waits for quorum restoration. This can be seen from the logs. However, I am unsure if it is called in a backoff manner.

I think it is because the default queue is https://github.com/openshift/library-go/blob/9991651fb2494e5aa03d1a82d782b7d398a7862c/vendor/k8s.io/client-go/util/workqueue/default_rate_limiters.go#L39

I have written a simple unit test to demonstrate my concern about a potential significant delay. The test simply requeues the same key in a loop, which is equivalent to returning [factory.SyntheticRequeueError](https://github.com/openshift/library-go/blob/02889d2d42d6e7e9c727bf18e029bf87a4c8a3d1/pkg/operator/revisioncontroller/revision_controller.go#L351) and having it requeued by the [factory](https://github.com/openshift/library-go/blob/fa4a6ca7a6ab8eb39259eda7b08ba2b7edb32a64/pkg/controller/factory/base_controller.go#L271), right?

The output from the test show that the same key will be delayed by `baseDelay*2^<num-failures>` where `baseDelay=5ms`:
```
=== RUN   TestQueueBackOff
Took: 5.677916ms to get an item off the queue for iteration: 0
Took: 11.042292ms to get an item off the queue for iteration: 1
Took: 21.045583ms to get an item off the queue for iteration: 2
Took: 41.051292ms to get an item off the queue for iteration: 3
Took: 81.067792ms to get an item off the queue for iteration: 4
Took: 161.067625ms to get an item off the queue for iteration: 5
Took: 321.077875ms to get an item off the queue for iteration: 6
Took: 641.077ms to get an item off the queue for iteration: 7
Took: 1.28104275s to get an item off the queue for iteration: 8
Took: 2.561077958s to get an item off the queue for iteration: 9
Took: 5.121045416s to get an item off the queue for iteration: 10
Took: 10.241192292s to get an item off the queue for iteration: 11
Took: 20.481249s to get an item off the queue for iteration: 12
Took: 40.961307125s to get an item off the queue for iteration: 13
Took: 1m21.92126675s to get an item off the queue for iteration: 14
```

Of course, this assumes that you would have to requeue the same key with no further updates to the queue, which is certainly possible. Given that restoring quorum might take a long time, the next time the queue receives an item, it could be significantly delayed, holding back the next rollout.

Does this make sense to you? Is this something we should be concerned about?

test:
```
func TestQueueBackOff(t *testing.T) {
	syncCtx := NewSyncContext("test", eventstesting.NewTestingEventRecorder(t))
	target := syncCtx.Queue()

	key := "key"
	for i := 0; i < 15; i++ {
		start := time.Now()
		target.AddRateLimited(key)
		target.Get()
		elapsed := time.Since(start)
		fmt.Println(fmt.Sprintf("Took: %v to get an item off the queue for iteration: %v", elapsed, i))
		target.Done(key)
	}
}
```

Reply:
Thank you @p0lyn0mial for getting back with the details.

To address the concern regarding the delay, can we revisit the PR [allow retry customization in syncContext](https://github.com/openshift/library-go/pull/1360). This would allow to customize the `DefaultControllerRateLimiter`, to have longer backoff periods.  In particular, by customizing the `SyncContext` in CEO, we could fine-tune the retry logic and reduce the delay, improving responsiveness.  

If this is acceptable can we address this in a follow-up PR and get the current changes in?

Reply:
How long does it take to repair the quorum ? Maybe we could just simply set a synthetic retry on the controller using https://github.com/openshift/library-go/blob/master/pkg/controller/factory/factory.go#L204 ?

Reply:
My understanding is that both the approaches would allow as to have longer backoff periods. But I am unsure which  would be more efficient. 

https://github.com/openshift/library-go/blob/b43f719e8c7ef2ee471af10cc29218593ee35de8/pkg/operator/revisioncontroller/revision_controller.go#L89

Would you suggest changing this to ```
		kubeInformersForTargetNamespace.Core().V1().Secrets().Informer()).WithSync(c.sync).ResyncEvery(resyncPeriod).ToController("RevisionController", eventRecorder) ```
		
If so, what would you suggest is an ideal `resyncPeriod`?

Reply:
no, I would add it directly to the `factory`, here https://github.com/openshift/library-go/blob/02889d2d42d6e7e9c727bf18e029bf87a4c8a3d1/pkg/operator/revisioncontroller/revision_controller.go#L80

For example:

```
return factory.New().WithInformers(
operatorClient.Informer(),
kubeInformersForTargetNamespace.Core().V1().ConfigMaps().Informer(),
kubeInformersForTargetNamespace.Core().V1().Secrets().Informer()).WithSync(c.sync).ResyncEvery(20*time.Second).ToController("RevisionController", eventRecorder)
```

Setting `ResyncEvery(20*time.Second)` instructs the factory to call the sync method every 20 seconds, regardless of the items in the queue. I believe this ensures that the precondition will be executed, even if requeuing was too aggressive and backed off items for several minutes.

I still would like to know how long does it take to repair the quorum ?

Reply:
> how long does it take to repair the quorum ?

On happy path, the repair process typically follows the restart loop of etcd, which usually takes about a minute or two. However, the recovery time can vary widely. In more complicated situations, it could be much longer — theoretically, you could run indefinitely without quorum as long as no write operations are occurring on the database. 

I have currently included the change in the PR to Resync every 20 seconds. Should a different interval be considered?